core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -99,43 +111,133 @@ class KubernetesClient(
+      .addNewContainer()
+      .withNewResources()
+      .withLimits(mapAsJavaMap(Map("memory" -> new Quantity(memory.toMB + "Mi"))))
+      .endResources()
+      .withName("user-action")
+      .withImage(image)
+      .withEnv(envVars)
+      .addNewPort()
+      .withContainerPort(8080)
+      .withName("action")
+      .endPort()
+      .endContainer()
+      .endSpec()
+      .build()
+
+    kubeRestClient.pods.inNamespace("openwhisk").create(pod)
 
-  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
     Future {
       blocking {
-        val pod =
-          kubeRestClient.pods().withName(id.asString).waitUntilReady(timeouts.inspect.length, timeouts.inspect.unit)
-        ContainerAddress(pod.getStatus().getPodIP())
+        val createdPod = kubeRestClient.pods
+          .inNamespace("openwhisk")
+          .withName(name)
+          .waitUntilReady(config.timeouts.run.length, config.timeouts.run.unit)
+        toContainer(createdPod)
       }
     }.recoverWith {
       case e =>
-        log.error(this, s"Failed to get IP of Pod '${id.asString}' within timeout: ${e.getClass} - ${e.getMessage}")
-        Future.failed(new Exception(s"Failed to get IP of Pod '${id.asString}'"))
+        log.error(this, s"Failed create pod for '$name': ${e.getClass} - ${e.getMessage}")
+        Future.failed(new Exception(s"Failed to create pod '$name'"))
     }
   }
 
-  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
-    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+  def rm(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
+    // Pod deletion will never complete if the pod contains a paused container.
+    // Therefore issue a resume before the delete (resuming a non-suspended container is harmless).
+    resume(container).map { _ =>
+      runCmd(Seq("delete", "--now", "pod", container.id.asString), config.timeouts.rm).map(_ => ())
+    }
+  }
 
-  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
-    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] = {
+    if (config.invokerAgent) {
+      Future {
+        blocking {
+          kubeRestClient
+            .inNamespace("openwhisk")
+            .pods()
+            .withLabel(key, value)
+            .list()
+            .getItems
+            .map { pod =>
+              rm(toContainer(pod))
+            }
+            .reduce((a, b) => a.flatMap(_ => b))
+        }
+      }
+    } else {
+      runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), config.timeouts.rm).map(_ => ())
+    }
+  }
 
-  def logs(id: ContainerId, sinceTime: Option[Instant], waitForSentinel: Boolean = false)(
+  def suspend(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
```

Comment:
Should we move those to the container instead of "faking" them in the client? Seems weird to pass the whole container to the client.

Reply:
I think the implementations belong in the KubernetesClient, since they need access to some of its state too (its config object). 

I agree passing the entire KubernetesContainer object as a parameter looks a bit backwards.  I decided to do that (instead of passing a few of its fields individually) as a hedge against future churn and because it felt like a nice logical unit (vs. passing containerId, workerIP, nativeContainerId).   But I don't feel that strongly about this.